storey/flow.py

```diff
@@ -608,73 +611,118 @@ async def _handle_completed(self, event, response):
-        else:
-            if len(self._batch) == 0:
-                self._batch_time = event.time
-                if self._timeout_secs:
-                    self._timeout_task = asyncio.get_running_loop().create_task(self._sleep_and_emit())
 
-            self._event_count = self._event_count + 1
-            self._batch.append(self._event_to_batch_entry(event))
+        key = self._get_event_key(event)
+
+        if len(self._batch[key]) == 0:
+            self._batch_time[key] = event.time
+
+        if self._timeout_secs and self._timeout_task is None:
+            self._timeout_task = asyncio.get_running_loop().create_task(self._sleep_and_emit())
 
-            if self._event_count == self._max_events:
-                if self._timeout_task and not self._timeout_task.cancelled():
-                    self._timeout_task.cancel()
-                await self._emit_batch()
+        self._event_count[key] = self._event_count[key] + 1
+        self._batch[key].append(self._event_to_batch_entry(event))
+
+        if self._event_count[key] == self._max_events:
+            if key == self._timeout_task_key and self._timeout_task and not self._timeout_task.cancelled():
+                self._timeout_task.cancel()
+                self._timeout_task_key = None
+            await self._emit_batch(key)
+
+    async def _sleep_and_emit(self):
+        while self._batch:
+            key = next(iter(self._batch.keys()))
+            time_delta = datetime.now(timezone.utc) - self._batch_time[key]
+            delta_seconds = time_delta.total_seconds()
+            if delta_seconds < self._timeout_secs:
+                self._timeout_task_key = key
+                await asyncio.sleep(self._timeout_secs - delta_seconds)
+            await self._emit_batch(key)
+
+        self._timeout_task.cancel()
+        self._timeout_task = None
+        self._timeout_task_key = None
+
+    def _event_to_batch_entry(self, event):
+        return self._get_event_or_body(event)
+
+    def _get_event_key(self, event):
+        key = None
+        if self._group_by_key:
+            if self.key_extractor:
+                key = self.key_extractor(event.body)
+            else:
+                key = event.key
+        return key
 
-    async def _emit_batch(self):
-        if len(self._batch) > 0:
-            batch_to_emit = self._batch
-            batch_time = self._batch_time
-            self._batch = []
-            self._batch_time = None
-            self._event_count = 0
+    async def _emit_batch(self, batch_key: Optional[str] = None):
+        batch_to_emit = self._batch.pop(batch_key)
+        batch_time = self._batch_time.pop(batch_key)
+        del self._event_count[batch_key]
+        await self._emit(batch_to_emit, batch_time)
 
-            await self._emit(batch_to_emit, batch_time)
+    async def _emit_all(self):
+        batch_keys = list(self._batch.keys())
+        for batch_key in batch_keys:
```

Comment:
Just
```
for batch_key in self._batch:
```
does the trick without creating an intermediate list.

Reply:
Can't do that since `self._emit_batch(batch_key)` changes `self._batch`'s size